### PR TITLE
Annotate `celery/exceptions.py`

### PR DESCRIPTION
**Pull request by Kludex**
- Related to #7394

### CODE REVIEW

**Review by auvipy — CHANGES_REQUESTED** [review]
please rebase & would be great if some unit test coverage can be increased

**Comment by Kludex** [comment]
I don't want to add tests on those PRs, as I'm not including any functionally I don't think it makes sense (or is fair).

That being said... I'm willing to create a plan to increase the coverage of the whole project to 100%, analogous to the plan created to add annotations (progressive, and by file(s)). I did https://github.com/encode/uvicorn/issues/102 for uvicorn.

If you agree, I can create a similar issue, and setup for it as well (so the coverage doesn't drop, only goes up).

**Review by GautamDhiman — APPROVED** [review]
LGTM

File: celery/exceptions.py [hunk 1]
```diff
@@ -144,17 +153,18 @@ class Retry(TaskPredicate):
...
     #: Exception (if any) that caused the retry to happen.
-    exc = None
+    exc: Optional[Exception] = None
+    excs: Optional[str] = None
 
-    #: Time of retry (ETA), either :class:`numbers.Real` or
-    #: :class:`~datetime.datetime`.
-    when = None
+    #: Time of retry (ETA)
+    when: Optional[Union[datetime, numbers.Number]] = None
```
[the remark this answers is not in the document]
Comment (Nusnus) [reply celery/exceptions.py:163]: I think you can leave it like that, but keep the original doc with the :class:`numbers.Real` doc, being a more specialized doc compared to the "general" annotation as `numbers.Number`.

What do you think?

**Review by Nusnus — APPROVED** [review]
Left one comment but not something critical to change, food for thought.
Besides, LGTM too.

File: celery/exceptions.py [hunk 2]
```diff
@@ -144,17 +153,18 @@ class Retry(TaskPredicate):
...
 
     #: Exception (if any) that caused the retry to happen.
-    exc = None
+    exc: Optional[Exception] = None
+    excs: Optional[str] = None
 
-    #: Time of retry (ETA), either :class:`numbers.Real` or
-    #: :class:`~datetime.datetime`.
-    when = None
+    #: Time of retry (ETA)
```
Comment (Kludex) [inline celery/exceptions.py:162]:
```suggestion
    #: Time of retry (ETA), either :class:`numbers.Number` or
    #: :class:`~datetime.datetime`.
```
@Nusnus  I didn't get what you said. Is this enough?

Comment (auvipy) [reply celery/exceptions.py:162]: :class:numbers.Real is what he suggested as per my understanding

Comment (Kludex) [reply celery/exceptions.py:162]: So... A mismatch between doc and code? 🧐

Comment (Nusnus) [reply celery/exceptions.py:162]: TBH you're right. My original thought was that the doc would mention Real and the annotation would mention Number, being _upper in hierarchy_, but that is wrong as you are noting @Kludex  :)

**Pull request closed without merging**